File: backend/elixir-phoenix/lib/rocket/ai/provider.ex
defmodule Rocket.AI.Provider do
  @moduledoc "OpenAI-compatible chat completions client using Req."

  alias Rocket.Engine.AppError

  defstruct [:base_url, :api_key, :model]

  @doc "Creates a new AI provider. Returns nil if not configured."
  def new(base_url, api_key, model) do
    if base_url && base_url != "" && api_key && api_key != "" && model && model != "" do
      %__MODULE__{
        base_url: String.trim_trailing(base_url, "/"),
        api_key: api_key,
        model: model
      }
    else
      nil
    end
  end

  @doc "Sends a system + user prompt to the LLM and returns the raw response text."
  def generate(%__MODULE__{} = provider, system_prompt, user_prompt) do
    url = "#{provider.base_url}/chat/completions"

    body = %{
      "model" => provider.model,
      "temperature" => 0.3,
      "response_format" => %{"type" => "json_object"},
      "messages" => [
        %{"role" => "system", "content" => system_prompt},
        %{"role" => "user", "content" => user_prompt}
      ]
    }

    case Req.post(url,
           json: body,
           headers: [
             {"content-type", "application/json"},
             {"authorization", "Bearer #{provider.api_key}"}
           ],
           receive_timeout: 120_000
         ) do
      {:ok, %Req.Response{status: status, body: resp_body}} when status >= 200 and status < 300 ->
        content =
          case resp_body do
            %{"choices" => [%{"message" => %{"content" => c}} | _]} when is_binary(c) and c != "" -> c
            _ -> nil
          end

        if content do
          {:ok, content}
        else
          {:error, AppError.new("AI_REQUEST_FAILED", 502, "AI provider returned empty response")}
        end

      {:ok, %Req.Response{status: status, body: resp_body}} ->
        detail =
          case resp_body do
            %{"error" => %{"message" => msg}} when is_binary(msg) -> msg
            body when is_binary(body) -> body
            body -> inspect(body)
          end

        {:error, AppError.new("AI_REQUEST_FAILED", 502, "AI provider returned #{status}: #{detail}")}

      {:error, reason} ->
        {:error, AppError.new("AI_REQUEST_FAILED", 502, "Failed to connect to AI provider: #{inspect(reason)}")}
    end
  end
end

File: backend/golang/internal/ai/provider.go
package ai

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"

	"rocket-backend/internal/engine"
)

// Provider is an OpenAI-compatible chat completions client.
type Provider struct {
	baseURL string
	apiKey  string
	model   string
	client  *http.Client
}

// NewProvider creates a new AI provider. Returns nil if not configured.
func NewProvider(baseURL, apiKey, model string) *Provider {
	if baseURL == "" || apiKey == "" || model == "" {
		return nil
	}
	return &Provider{
		baseURL: strings.TrimRight(baseURL, "/"),
		apiKey:  apiKey,
		model:   model,
		client:  &http.Client{},
	}
}

// Model returns the configured model name.
func (p *Provider) Model() string {
	return p.model
}

type chatRequest struct {
	Model          string        `json:"model"`
	Temperature    float64       `json:"temperature"`
	ResponseFormat responseFmt   `json:"response_format"`
	Messages       []chatMessage `json:"messages"`
}

type responseFmt struct {
	Type string `json:"type"`
}

type chatMessage struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

type chatResponse struct {
	Choices []struct {
		Message struct {
			Content string `json:"content"`
		} `json:"message"`
	} `json:"choices"`
}

type apiError struct {
	Error struct {
		Message string `json:"message"`
	} `json:"error"`
}

// Generate sends a system + user prompt to the LLM and returns the raw response text.
func (p *Provider) Generate(systemPrompt, userPrompt string) (string, error) {
	url := p.baseURL + "/chat/completions"

	body := chatRequest{
		Model:          p.model,
		Temperature:    0.3,
		ResponseFormat: responseFmt{Type: "json_object"},
		Messages: []chatMessage{
			{Role: "system", Content: systemPrompt},
			{Role: "user", Content: userPrompt},
		},
	}

	payload, err := json.Marshal(body)
	if err != nil {
		return "", engine.NewAppError("AI_REQUEST_FAILED", 502, "Failed to marshal AI request")
	}

	req, err := http.NewRequest("POST", url, bytes.NewReader(payload))
	if err != nil {
		return "", engine.NewAppError("AI_REQUEST_FAILED", 502, "Failed to create AI request")
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+p.apiKey)

	resp, err := p.client.Do(req)
	if err != nil {
		return "", engine.NewAppError("AI_REQUEST_FAILED", 502,
			fmt.Sprintf("Failed to connect to AI provider: %v", err))
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", engine.NewAppError("AI_REQUEST_FAILED", 502, "Failed to read AI response")
	}

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		var apiErr apiError
		detail := string(respBody)
		if json.Unmarshal(respBody, &apiErr) == nil && apiErr.Error.Message != "" {
			detail = apiErr.Error.Message
		}
		return "", engine.NewAppError("AI_REQUEST_FAILED", 502,
			fmt.Sprintf("AI provider returned %d: %s", resp.StatusCode, detail))
	}

	var chatResp chatResponse
	if err := json.Unmarshal(respBody, &chatResp); err != nil {
		return "", engine.NewAppError("AI_REQUEST_FAILED", 502, "Failed to parse AI response")
	}

	if len(chatResp.Choices) == 0 || chatResp.Choices[0].Message.Content == "" {
		return "", engine.NewAppError("AI_REQUEST_FAILED", 502, "AI provider returned empty response")
	}

	return chatResp.Choices[0].Message.Content, nil
}
